### ISSUE TITLE
Conditions of PDA Acceptance

### ISSUE THREAD

**Issue by YtvwlD** [caleb531/automata#8]
DPDAs can accept when the stack is empty or a final state has been reached (assuming that all input has been read).

The related code is `DPDA._check_for_input_rejection` which is basically this:
```python
if current_state not in self.final_states and stack:
            raise exceptions.RejectionException()
```

This leads to the problem that when the DPDA accepts via final state, it will also accept if the stack is empty after reading a given input, even when no final state has been reached.

I would prefer specifying whether a DPDA accepts either via empty stack or via final state.

(This can probably be deducted by looking at whether `self.final_states` is empty but I'm not so sure about that.)

**Comment by caleb531** [comment]
@YtvwlD Any update on how you'd like to handle PDA acceptance? I'd like to get Automata v3 released soon, but would also like to hear your input first in case any more breaking API changes need to be made.

**Comment by YtvwlD** [comment]
I'm really sorry to not have responded earlier. I really hope to get to this next week.

**Comment by caleb531** [comment]
It's quite alright, Niklas! :) I look forward to hearing your response when you can get to it.

Caleb

**Comment by caleb531** [comment]
Hey, @YtvwlD! I just wanted to check in if you happened to have some time this week to comment more on this issue. If not, that's alright—I'm just checking. :)

**Comment by YtvwlD** [comment]
No, sorry. But I hope to be able to take a look before the year ends.

**Comment by caleb531** [comment]
@YtvwlD That's fine! :)

**Comment by caleb531** [comment]
@YtvwlD I'm eager to start writing release notes and get Automata v3 published since more people are finding this library (no doubt thanks to your efforts in making it more comprehensively useful). I am still interested in hearing your feedback, but will require now that any changes you recommend be backwards-compatible with the pending v3 API (because I am not interested in pushing a v4 for the sake of this one behavioral change).

That being said, I have done some research on the topic myself and have concluded that there's no hard rule mandating that a given PDA cannot accept by both modes simultaneously. Besides, if I recall correctly, one can simply add a lambda/epsilon transition at the start of an automaton to avoid the automatic empty-state acceptance.

Ultimately, I'd still consider an additional parameter for the user to customize the behavior on-the-fly, but I am definitely committed to keeping the current behavior as the default. I will keep this issue open for the sake of discussion and because this is still under consideration.

**Comment by YtvwlD** [comment]
Sorry for not getting to this earlier.

From what I've learnt (and which is stated in the [German Wikipedia Article](https://de.wikipedia.org/wiki/Kellerautomat); I'm translating the relevant bit), (nondeterministic) PDAs are defined in a way that:

> […]
> When the whole input has been read and the stack is empty, the input is valid.
>
> There are a few different definitions:
> * The language can be defined as the set of words that cause the PDA to reach a so-called final state, with no regard to whether the stack is empty.
> […]

The [English Wikipedia article](https://en.wikipedia.org/wiki/Pushdown_automaton) confused me a bit - especially the part you quoted above.

But there'a also a part below that: (Highlighting by me.)

> The pushdown automaton either accepts by final state, which means after reading its input the automaton reaches an accepting state (in F), or it accepts by empty stack ( ε ), which means after reading its input the automaton empties its stack. The first acceptance mode uses the internal memory (state), the second the external memory (stack).
> 
> Formally one defines
> 
>    L ( M ) = { w ∈ Σ ∗ | ( q 0 , w , Z ) ⊢ M ∗ ( f , ε , γ ) with f ∈ F and γ ∈ Γ ∗ } (final state)
>    N ( M ) = { w ∈ Σ ∗ | ( q 0 , w , Z ) ⊢ M ∗ ( q , ε , ε ) with q ∈ Q } (empty stack)
> […]
> For each single pushdown automaton these two languages need to have no relation: they may be equal but usually this is not the case. **A specification of the automaton should also include the intended mode of acceptance.** Taken over all pushdown automata both acceptance conditions define the same family of languages.

I think this is similar to my initial proposal:

> I would prefer specifying whether a DPDA accepts either via empty stack or via final state.

(This whole thing seems to be unimportant, because NPDAs accepting on final state and those accepting on an empty stack can do the same things. But this is not true for DPDAs.)

**Comment by caleb531** [comment]
Fair enough. Maybe the parameter can be called something like `acceptance_mode`? Perhaps values can be either `'final_state'` or `'empty_stack'` or `'all'`/`'both'`. I'm open to suggestions if you have different ideas for the names.

I'll consider any PR you send my way. :)

**Comment by YtvwlD** [comment]
Yes, this looks good. :)

**Comment by YtvwlD** [comment]
I created PR #12 for this.

**Comment by caleb531** [comment]
@YtvwlD I've just merged $12 with the new functionality. Thanks for the hard work putting that together.

**Comment by YtvwlD** [comment]
Thanks. :)

**Comment by caleb531** [comment]
@YtvwlD Just wanted to let you know that I've officially released Automata v3.1 with the new `acceptance_mode` param. :)
